notification-server/src/database.rs

```diff
@@ -0,0 +1,207 @@
+use crate::models::Preference;
+use anyhow::{anyhow, Context};
+use deadpool_postgres::{Manager, ManagerConfig, Pool, RecyclingMethod};
+use lazy_static::lazy_static;
+use std::collections::{HashMap, HashSet};
+use tokio_postgres::NoTls;
+
+#[derive(Clone, Debug)]
+pub struct PreparedStatements {
+    get_devices_from_account: tokio_postgres::Statement,
+    upsert_device:            tokio_postgres::Statement,
+    pool:                     Pool,
+}
+
+impl PreparedStatements {
+    async fn new(pool: Pool) -> anyhow::Result<Self> {
+        let mut client = pool.get().await.context("Failed to get client")?;
+        let transaction = client
+            .transaction()
+            .await
+            .context("Failed to start a transaction")?;
+        let get_devices_from_account = transaction
+            .prepare("SELECT device_id FROM account_device_mapping WHERE address = $1 LIMIT 1000")
+            .await
+            .context("Failed to create account device mapping")?;
+        let upsert_device = transaction
+            .prepare(
+                "INSERT INTO account_device_mapping (address, device_id, preferences) VALUES ($1, \
+                 $2, $3) ON CONFLICT (address, device_id) DO UPDATE SET preferences = \
+                 EXCLUDED.preferences;",
+            )
+            .await
+            .context("Failed to create account device mapping")?;
+        transaction
+            .commit()
+            .await
+            .context("Failed to commit transaction")?;
+        Ok(PreparedStatements {
+            get_devices_from_account,
+            upsert_device,
+            pool,
+        })
+    }
+
+    pub async fn get_devices_from_account(&self, address: &[u8]) -> anyhow::Result<Vec<String>> {
+        let client = self
+            .pool
+            .get()
+            .await
+            .map_err(|e| anyhow!("Failed to get client: {}", e))?;
+        let params: &[&(dyn tokio_postgres::types::ToSql + Sync)] = &[&address];
+        let rows = client.query(&self.get_devices_from_account, params).await?;
+        let devices: Vec<String> = rows
+            .iter()
+            .map(|row| row.try_get::<_, String>(0))
+            .collect::<Result<Vec<String>, _>>()?;
+        Ok(devices)
+    }
+
+    pub async fn upsert_subscription(
+        &self,
+        address: Vec<Vec<u8>>,
+        preferences: Vec<Preference>,
+        device_id: &str,
+    ) -> anyhow::Result<()> {
+        let mut client = self
+            .pool
+            .get()
+            .await
+            .map_err(|e| anyhow!("Failed to get client: {}", e))?;
```

Comment:
Maybe too much for this PR, but for follow-up PRs: Since this error happens several times throughout the code whenever you request the pool. As such there is usually a `enum Error` that makes it easier to re-use the Error variant.